### PR TITLE
 bpo-45459: Add Py_buffer to limited API

### PR DESCRIPTION
**Pull request by tiran**
- [x] ``Py_buffer`` struct
- [x] ``PyBuffer_*()`` API functions
- [x] ``PyBUF_*`` constants
- [x] ``Py_bf_getbuffer`` and ``Py_bf_releasebuffer`` type slots
- [x] ``PyMemoryView_FromBuffer()`` API
- [x] tests for limited API
- [x] ``make regen-limited-abi``
- [x] documentation update
- [ ] export ``PyPickleBuffer*()`` API ???

https://bugs.python.org/issue45459

Automerge-Triggered-By: GH:encukou

### CODE REVIEW

**Comment by tiran** [comment]
@encukou could you take a look, please?

**Comment by encukou** [comment]
I'll get to it this week.

File: Doc/data/stable_abi.dat [hunk 1]
```diff
@@ -476,8 +485,10 @@ function,PyObject_CallMethodObjArgs,3.2,
 function,PyObject_CallNoArgs,3.10,
 function,PyObject_CallObject,3.2,
 function,PyObject_Calloc,3.7,
+function,PyObject_CheckBuffer,3.11,
 function,PyObject_CheckReadBuffer,3.2,
 function,PyObject_ClearWeakRefs,3.2,
+function,PyObject_CopyData,3.11,
```
Comment (encukou) [inline Doc/data/stable_abi.dat:491]: This function is undocumented, can you also add it to the docs? (Or exclude it from the limited API)

File: Include/buffer.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+/* === New Buffer API ============================================
+ * Limited API and stable ABI since Python 3.11
+ *
+ * Py_buffer struct layout and size is now part of the stable abi3. The
+ * struct layout and size must not be changed in any way, as it would
+ * break the ABI.
+ *
+ */
+
+typedef struct bufferinfo {
```
Comment (encukou) [inline Include/buffer.h]: Can we leave out the `bufferinfo` tag? Anonymous structs are well-supported nowadays, and only `Py_buffer` should be public.

File: Include/cpython/object.h [hunk 3]
```diff
@@ -45,61 +45,14 @@ typedef struct _Py_Identifier {
...
-    int ndim;
-    char *format;
-    Py_ssize_t *shape;
-    Py_ssize_t *strides;
-    Py_ssize_t *suboffsets;
-    void *internal;
-} Py_buffer;
-
+/* Py_buffer is defined in buffer.h */
+typedef struct bufferinfo Py_buffer;
```
Comment (encukou) [inline Include/cpython/object.h]: This shouldn't be necessary, the API should be included from Python.h.

File: Modules/xxlimited.c [hunk 4]
```diff
@@ -19,6 +19,8 @@
           def __init__(self):
               # In the C class, "_x_attr" is not accessible from Python code
               self._x_attr = {}
+              self._x_buffer = bytesarray(10)
+              self._x_exports = 0
```
Comment (encukou) [inline Modules/xxlimited.c]: IMO, it's better to explain this in English than in Python.
The name `_x_buffer` isn't used elsewhere, so there's not much point in introducing it.
```suggestion
```

File: Modules/xxlimited.c [hunk 5]
```diff
@@ -29,6 +31,10 @@
           def __delattr__(self, name):
               del self._x_attr[name]
 
+          @property
+          def x_exports(self):
+              return self._x_exports
+
```
Comment (encukou) [inline Modules/xxlimited.c]:
```suggestion
          @property
          def x_exports(self):
              """Return the number of times an internal buffer is exported."""
              # Each Xxo instance has a 10-byte buffer that can be
              # accessed via the buffer interface (e.g. `memoryview`).
              return self._x_exports
```

File: Doc/data/stable_abi.dat (hunk 1 above)
Comment (tiran) [reply Doc/data/stable_abi.dat:491]: I decided to document the function. It's used by Cython.

File: Include/buffer.h (hunk 2 above)
Comment (tiran) [reply Include/buffer.h]: Good idea!

File: Include/cpython/object.h (hunk 3 above)
Comment (tiran) [reply Include/cpython/object.h]: There is a chicken and egg problem. ``buffer.h`` must be included after ``object.h`` has declared ``PyObject``. It's now included twice, one time by ``Python.h`` for public API and one time by ``cpython/object.h`` for internal use.

**Review by encukou — APPROVED** [review]
Looks good, let's get it in! Thanks!

**Comment by rdb** [comment]
This change broke our project build because when cpython/object.h is including buffer.h it is forcing it to resolve along the search path, and the compiler is hitting the buffer.h in our project rather than the one in the Python include directory.

Should it not be using a relative include, ie. `#include "../buffer.h"` ?  I think otherwise this change will cause breakage for many projects given how common the header name "buffer.h" may be.

**Pull request merged**